### ISSUE TITLE
[Bug][fn] Go functions do not reset idleTimer correctly.

### ISSUE THREAD

**Issue by flowchartsman** [apache/pulsar#20449]
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

The code in `instance.go` which handles incoming messages and idle timeout basically boils down to this:

```go
	idleDuration := getIdleTimeout(time.Millisecond * time.Duration(gi.context.instanceConf.killAfterIdleMs))
	idleTimer := time.NewTimer(idleDuration)
	defer idleTimer.Stop()

CLOSE:
	for {
		idleTimer.Reset(idleDuration)
		select {
		case cm := <-channel:
			// handle message
		case <-idleTimer.C:
			close(channel) // this is not great either, but that's a separate bug ;)
			break CLOSE
		}
	}
```
However, this does not account for the idle timer firing during work. If it does, then the reset will not have the desired effect, and the idle clause will fire erroneously on a subsequent loop, since there's now a timestamp on the channel.

### Minimal reproduce step

Here's a simple repro:
```go
package main

import "fmt"
import "time"

func main() {
	const idleDuration = 250*time.Millisecond
	work := make(chan int,2)
	work <- 1
	work <- 2
	t := time.NewTimer(idleDuration)
	EVENTLOOP:
	for {
		t.Reset(idleDuration)
		fmt.Println("waiting for new work")
		select {
		case w := <- work:
			fmt.Printf("doing work on %d\n", w)
			time.Sleep(500*time.Millisecond)
			fmt.Println("done with work")
		case <- t.C:
			fmt.Println("timed out")
			break EVENTLOOP
		}
	}
}
```

Run that a few times, and you'll see:
```
waiting for new work
doing work on 1
done with work
waiting for new work
timed out
```
This is exactly what would happen in a Go function, only it would then call `log.Fatal`

### What did you expect to see?

n/a

### What did you see instead?

n/a

### Anything else?

The correct fix is to call `Stop()` on the timer and, if it returns false, drain the channel before resetting:

```go
CLOSE:
	for {
		select {
		case cm := <-channel:
			// handle message
		case <-idleTimer.C:
			close(channel)
			break CLOSE
		}
		if !idleTimer.Stop() {
			<-idleTimer.C
		}
	}

```

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

**Comment by lhotari** [comment]
@flowchartsman Thanks for reporting and contributing the fix for this issue.

I recently happened to review some code in the Pulsar Go client and it contains similar code for handling the send timeout on the producer. This is the code: https://github.com/apache/pulsar-client-go/blob/2eb340511dbf0602f7c561a1d86e179098f4662b/pulsar/producer_partition.go#L910-L924
Does this have a similar issue?

**Comment by flowchartsman** [comment]
> @flowchartsman Thanks for reporting and contributing the fix for this issue.
> 
> I recently happened to review some code in the Pulsar Go client and it contains similar code for handling the send timeout on the producer. This is the code: https://github.com/apache/pulsar-client-go/blob/2eb340511dbf0602f7c561a1d86e179098f4662b/pulsar/producer_partition.go#L910-L924 Does this have a similar issue?

From a quick glance, I'd say this looks fine. The loop is a range over the channel, which guarantees that it will be read before `Reset` is called, and each use of `Reset` is immediately followed by a control flow statement that begins the next loop iteration, so none of the timeout is expended before waiting on it again. Finally, on the off chance that the timer fires after the loop is exited, the deferred `Stop` should ensure that the buffered `C` channel is GCed.

This avoids the main problem we had here where the timer wasn't guaranteed to be read in each loop, meaning it could fire during work and be left until the next iteration.
